File: 2/2/pdoo/prac/ruby/damage.rb
#encoding:utf-8

require "./lib/DamageToUI.rb"
require_relative "./weapontype.rb"
require_relative "./weapon.rb" # No creo que sea necesario

module Deepspace
  # Daño
  class Damage

    attr_reader :nWeapons, :weapons, :nShields

    def initialize(w,wl,s)
      @nWeapons = w
      @weapons = wl
      @nShields = s
    end
    
    # Constructor de daño con número genérico de armas
    def self.newNumericWeapons(w,s)
      new(w,nil,s)
    end

    # Constructor de daño con tipo específico de armas
    def self.newSpecificWeapons(wl,s)
      new(-1,wl,s)
    end

    def getUIversion
      DamageToUI.new(self)
    end
    
    def to_s
      getUIversion.to_s
    end

    # Comprueba si el array 'w' contiene armas del tipo 't'
    private def arrayContainsType(w,t)
      w.each_index do |i|
        if w[i].type==t
          return i
        end
      end
      return -1
    end

    # Ajusta el objeto al equipo del jugador
    def adjust(w,s)
      newNShields = [s.length,@nShields].min
      if @nWeapons==-1
        newWeapons = Array.new(@weapons)
        newWeapons.each do |type|
          if arrayContainsType(w,type)==-1
            newWeapons.delete(type)
          end
        end
        return self.class.newSpecificWeapons(newWeapons,newNShields)
      elsif @weapons==nil
        newNWeapons = [w.length,@nWeapons].min
        return self.class.newNumericWeapons(newNWeapons,newNShields)
      end
    end

    # Elimina un arma del daño
    def discardWeapon(w)
      if @weapons!=nil
        index = arrayContainsType(@weapons,w.type)
        if index!=-1
          @weapons.delete_at(index)
        end
      elsif @nWeapons>0
        @nWeapons -= 1
      end
    end

    # Elimina un potenciador de escudo
    def discardShieldBooster
      if @nShields>0
        @nShields -= 1
      end
    end

    # Comprueba si el daño no tiene efecto
    def hasNoEffect
      if @nShields==0
        if @weapons!=nil and @weapons.length==0
          return true
        elsif @nWeapons==0
          return true
        else
          return false
        end
      end
      return false
    end

    private_class_method :new

  end
end

File: 2/2/pdoo/prac/java/deepspace/src/Deepspace/Damage.java
/**
 * @author Manuel Gachs Ballegeer
 */
package Deepspace;

import java.util.ArrayList;

/**
 * @brief Clase del daño
 */
class Damage {
    private int nShields;
    private int nWeapons;
    private ArrayList<WeaponType> weapons = new ArrayList<>();
    
    Damage(int w, int s) {
        nWeapons = w;
        nShields = s;
        weapons = null;
    }
    
    Damage(ArrayList<WeaponType> wl,int s) {
        nWeapons = -1;
        weapons = wl;
        nShields = s;
    }
    
    Damage(Damage d) {
        nWeapons = d.nWeapons;
        nShields = d.nShields;
        weapons = d.weapons;
    }
    
    DamageToUI getUIversion() { return new DamageToUI(this); }
    
    private int arrayContainsType(ArrayList<Weapon> w,WeaponType t) {
        for (int i=0;i<w.size();i++) {
            if (w.get(i).getType()==t)
                return i;
        }
        return -1;
    }
    
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        // Ajustamos los escudos
        int newNShields = (s.size()<=nShields)?s.size():nShields;
        // Ajustamos las armas
        if (weapons==null) {
            int newNWeapons = (w.size()<=nWeapons)?w.size():nWeapons;
            return new Damage(newNWeapons,newNShields);
        } else if (nWeapons==-1) {
            ArrayList<WeaponType> newWeapons = new ArrayList<>();
            for (int i=0;i<weapons.size();i++) {
                int index_aux = arrayContainsType(w,weapons.get(i));
                if (index_aux!=-1) {
                    int cantidad_wt = 0, cantidad_w = 0;
                    for (int j=i;i<weapons.size();i++)
                        if (weapons.get(j)==weapons.get(i))
                            cantidad_wt++;
                    for (Weapon weap: w)
                        if (weap.getType()==weapons.get(i))
                            cantidad_w++;
                    if (cantidad_wt<=cantidad_w)
                        newWeapons.add(weapons.get(i));
                 }
            }
            return new Damage(newWeapons,newNShields);
        } else
            return null;
    }
    
    public void discardWeapon(Weapon w) {
        if (weapons==null) {
            if (nWeapons<0)
                nWeapons--;
        } else {
            boolean discarded = false;
            for (int i=0;i<weapons.size() || !discarded;i++) 
                if (weapons.get(i)==w.getType()) {
                    weapons.remove(i);
                    discarded = true;
                }
        }
    }
    
    public void discardShieldBooster() {
        if (nShields>0)
            nShields--;
    }
    
    public boolean hasNoEffect() {
        if (nShields==0) {
            return weapons.isEmpty() && (nWeapons==0 || nWeapons==-1);
        }
        return false;
    }
    
    public int getNShields() { return nShields; }
    
    public int getNWeapons() { return nWeapons; }
    
    public ArrayList<WeaponType> getWeapons() { return weapons; }
    
    @Override
    public String toString() {
        String aux = "Número de escudos: "+nShields;
        aux += "\nNúmero de armas: ";
        if (nWeapons!=-1)
            aux += nWeapons;
        aux += "\nTipos de armas:";
        if (weapons!=null)
            for (WeaponType t: weapons)
                aux += " "+t+",";
        return aux;
    }
}
